
<!DOCTYPE html>
<html>
<head>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  <title>Blood eQTL browser</title>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">


    <script src="https://code.jquery.com/jquery-1.12.4.min.js"></script>
    <script src="https://code.jquery.com/ui/1.12.1/jquery-ui.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-mousewheel/3.1.13/jquery.mousewheel.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/underscore.js/1.9.1/underscore-min.js"></script>


    <script src="{{ interfix }}{{ url_for('static', filename='promise.js') }}"></script>
    <script src="{{ interfix }}{{ url_for('static', filename='zlib_and_gzip.js') }}"></script>
    <script src="{{ interfix }}{{ url_for('static', filename='inflate.js') }}"></script>


    <link rel="stylesheet" type="text/css" href="{{ interfix }}{{ url_for('static', filename='igv.css') }}"/>
    <link rel="stylesheet" type="text/css" href="{{ interfix }}{{ url_for('static', filename='fa-svg-with-js.css') }}"/>
    <script type="text/javascript" src="{{ interfix }}{{ url_for('static', filename='igv.min.js') }}"></script>

	<script type="text/javascript" src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
	<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <link rel="stylesheet" type="text/css" href="https://rawgit.com/wenzhixin/bootstrap-table/master/src/bootstrap-table.css">
    <script type="text/javascript" src="https://rawgit.com/wenzhixin/bootstrap-table/master/src/bootstrap-table.js"></script>
    <script src="https://rawgit.com/wenzhixin/bootstrap-table/master/src/extensions/filter-control/bootstrap-table-filter-control.js"></script>


  <style type="text/css">
    #page-wrapper {
    padding: 0 2.6%;
    padding-top: 80px;
    padding-right: 2.6%;
    padding-bottom: 0px;
    padding-left: 2.6%;
    position: relative;
	}



  </style>

  <script type="text/javascript">
  	window.eqtl_detail_tracks = [];

	$(window).load(function(){
		$(function () {
		});
	    $('#lead_table').on('post-body.bs.table', function (e, data) {
	        //set_header_tt(e.target.getAttribute("id"));
	        $('[data-toggle="tooltip"]').tooltip();
	    });
    });

	////// To highlight individual rows
	function rowStyle(row, index) {
	  if (window.eqtl_detail_tracks.indexOf(row.id) != -1) {
	    return {
	      classes: 'text-nowrap another-class',
	      css: {"background-color": "#b2ad7f"}
	    };
	  } else {
	    if (row.hasOwnProperty("my_hidden_id") && row["my_hidden_id"] == "gajhsdg7"){
	      return {css: {"color": "green", "font-size": "15px"}}
	    }
	    return {};
	  }

	}

    function lead_fn_formatter(value, row, index, field) {
        glyphs = {"glyphicon-search": 'igv_go_to(\'chr'+row.chromosome+'\', '+row.start+')', "glyphicon-plus": 'create_track_by_id(\''+row.id+'\', \''+row.geneSymbol+'\')', "glyphicon-eye-open": 'add_scatter(\''+row.id+'\')'}
        glyph_tooltips = {"glyphicon-search": 'Zoom in genome browser', "glyphicon-plus": 'Add to genome browser', "glyphicon-eye-open": 'Display expression boxplot'}
        html_str = "<div>"
        for (gl in glyphs) {
        	if ((gl == 'glyphicon-plus') && (window.eqtl_detail_tracks.indexOf(row.id) != -1)) continue;
            html_str += '<a href="#genome_browser"><span class="glyphicon '+gl+'" aria-hidden="true" onclick="'+glyphs[gl]+'" data-toggle="tooltip" data-placement="bottom" title="'+glyph_tooltips[gl]+'"></span></a> ';
        }
        html_str += "</div>"
        return html_str
    }


   function add_scatter(id) {
        href = "http://127.0.0.1:8089/expr_boxplot?id="+id;
        window.open(href);
    }
    
    function round_nicely(value, row, index, field){
        return value.toPrecision(3);
    }

    function refresh_table(table_id){
        $('#'+table_id).bootstrapTable('refresh', {silent: true});
        //$('#'+table_id).bootstrapTable('refresh');
    }

    function set_header_tt(table_id){
        ranges_tt =  "Include '>' or '<' to search in ranges";
        header_tooltips = {'beta': ranges_tt, 'pValue': ranges_tt};
        for (df in header_tooltips){
            $('#'+table_id+'.th').find('[data-field="'+df+'"]').attr('data-toggle', 'tooltip').attr('data-placement', 'bottom').attr('title', header_tooltips[df]);
        }
    }




</script>

    <script type="text/javascript">

    <!-- TODO:FIX THIS -->
    igv_eqtl_baseurl = 'http://127.0.0.1:8089/';
    table_ajax_url = igv_eqtl_baseurl;

    refresh_eqtl_detail_tracks = function(){
    	all_tracks = igv.browser.trackViews;
		var tracks_ids = [];
		all_tracks.forEach(function (trackView) {
		    if (trackView.track instanceof igv.BloodSignalEqtlTrack){
		        if (trackView.track.config.id) {
					tracks_ids.push(trackView.track.config.id);
		        }
		    }
		})
		window.eqtl_detail_tracks = tracks_ids;
		refresh_table('lead_table');
    }


    create_lead_track = function(cellType, dataset){
        igv.browser.loadTrack({
                         type: 'bloodeqtl',
                         sourceType: 'bloodeqtl',
                         url: igv_eqtl_baseurl,
                         dataset: dataset,
                         cellType: cellType,
                         name: cellType + ' blood eQTL data',
                         });
    };


     create_track= function(variant, probeId, fwdIter, cellType, dataset, hgnc, var_chrom="", var_pos=0){
        igv.browser.loadTrack({
                         type: 'bloodsignaleqtl',
                         sourceType: 'bloodsignaleqtl',
                         url: igv_eqtl_baseurl,
                         queryVarId:variant,
                         queryProbeId:probeId,
                         fwdIter:fwdIter,
                         cellType: cellType,
                         name: cellType+': ' + hgnc + '(' + probeId + '): ' + variant,
                         dataset: dataset,
                         id: 'blablabla',
                         });
        if ((var_chrom != "") && (var_pos != 0)){
            igv.browser.goto(var_chrom, var_pos-10000, var_pos+10000);
        }

    };

    create_track_by_id= function(id, hgnc = "", var_chrom="", var_pos=0){
    	// The name generation could be done a bit more elegantly with another query to the server 
    	var id_splits = id.split("_"),
    		cellType = id_splits[2],
    		probeId = id_splits[3],
    		variant = id_splits[4];
        igv.browser.loadTrack({
                         type: 'bloodsignaleqtl',
                         sourceType: 'bloodsignaleqtl',
                         url: igv_eqtl_baseurl,
                         id:id,
                         name: cellType+': ' + hgnc + ' (' + probeId + '): ' + variant,
                         });
        if ((var_chrom != "") && (var_pos != 0)){
            igv.browser.goto(var_chrom, var_pos-10000, var_pos+10000);
        }

    };

    igv_go_to = function(var_chrom, var_pos){
    	if ((var_chrom != "") && (var_pos != 0)){
            igv.browser.goto(var_chrom, var_pos-10000, var_pos+10000);
        }
    }


    create_epi_tracks = function(cellType){
        $.getJSON( igv_eqtl_baseurl + 'get_epi_tracks?cellType='+cellType, function( data ) {

            for (i in data){
                igv.browser.loadTrack(data[i]);
            };
          //$.each( data, function( inner_dict ) {
          //  config = {name: inner_dict.name, type: inner_dict.type, height: inner_dict.height, url: inner_dict.url};
          //  igv.browser.loadTrack(config);
          //});
        });

    };

    </script>



    <script type="text/javascript">

    $(document).ready(function () {

        var div,
                options,
                browser;

        div = $("#igv_div")[0];
        options = {
            genome: "hg19",
            showNavigation: true,
            fastaURL: 'https://s3.amazonaws.com/igv.broadinstitute.org/genomes/seq/hg19/hg19.fasta',
            cytobandURL: 'https://s3.amazonaws.com/igv.broadinstitute.org/genomes/seq/hg19/cytoBand.txt',
            // locus: 'SLC25A3',
            locus: 'rs28372744',
            // locus: ['egfr', 'myc', 'pten'],
            // locus: ['2', '4', '8'],
            flanking: 7500,
            //locus: 'SLC25A3',
            locus: 'chr1:241461541-249250621',
            // locus: ['egfr', 'myc', 'pten'],
            // locus: ['2', '4', '8'],
            formats: {
                gwasSNPS: {
                    coords: 0,
                    chr: 0,
                    start: 1,
                    end: 2,
                    fields: ["chrom", "chromStart", "chromEnd", "Strongest SNP-risk allele", "Disease/Phenotype", "P-value", "Odds ratio or beta", "PUBMEDID"]
                }
            },
            tracks: [
                {
                     type: 'bloodeqtl',
                     sourceType: 'bloodeqtl',
                     url: igv_eqtl_baseurl,
                     dataset: 'merged',
                     cellType: 'CD14',
                     name: 'CD14 blood eQTL data',
                },
                {
                    featureType: 'genes',
                    url: 'http://s3.amazonaws.com/igv.broadinstitute.org/annotations/hg19/genes/gencode.v18.collapsed.bed',
                    name: 'Genes',
                    order: 10000,
                    displayMode: 'EXPANDED',
                    height: 100,
                    autoHeight: true,
                    removable: false,
                    labelDisplayMode: 'SLANT'
                }
            ]
        };

        browser = igv.createBrowser(div, options);

        refresh_eqtl_detail_tracks();
        // igv.browser.search("psphp1");

        igv.browser.on("trackremoved", function(e){
			refresh_eqtl_detail_tracks();
        });

        igv.browser.on("trackadded", function(e){
			refresh_eqtl_detail_tracks();
        });

    });


    </script>



	<script type="text/javascript">

	function cancel_file_picker(evt){
	    new_file = document.getElementById("new_file");
	    new_idxfile = document.getElementById("new_idxfile");
	    files = [new_file, new_idxfile]
	    for (i in files){
	        try{
	            files[i].value = '';
	            if(files[i].value){
	                files[i].type = "text";
	                files[i].type = "file";
	            }
	        }catch(e){}
	    }
	    file_picker_modal.style.display = "none";
	}

	function add_file_picker(evt){
	    new_file = document.getElementById("new_file");
	    new_idxfile = document.getElementById("new_idxfile");

	    track_data = {url: new_file.files[0]};
	    if (new_idxfile.value != ''){
	        track_data['indexURL'] = new_idxfile.files[0];
	    }

	    igv.browser.loadTrack(track_data);
	    cancel_file_picker(evt);
	}

	$(document).ready(function () {

		    // Get the modal
		var file_picker_modal = document.getElementById('filePicker');

		// Get the button that opens the modal
		var file_picker_modal_btn = document.getElementById("filePickerModalBtn");

		// Get the <span> element that closes the modal
		var close_file_picker = document.getElementsByClassName("close-filePicker")[0];

		// When the user clicks on the button, open the modal
		file_picker_modal_btn.onclick = function() {
		    file_picker_modal.style.display = "block";
		}

		document.getElementById('filePickerCancelBtn').addEventListener('click', cancel_file_picker, false);
		document.getElementById('filePickerAddBtn').addEventListener('click', add_file_picker, false);

		// When the user clicks on <close_file_picker> (x), close the modal
		close_file_picker.onclick = function() {
		    file_picker_modal.style.display = "none";
		}

		// When the user clicks anywhere outside of the modal, close it
		window.onclick = function(event) {
		    if (event.target == file_picker_modal) {
		        file_picker_modal.style.display = "none";
		    }
		}
	});
	</script>

    <style type="text/css">
        /* The Modal (background) */
.modal {
    display: none; /* Hidden by default */
    position: fixed; /* Stay in place */
    z-index: 1001; /* Sit on top */
    left: 0;
    top: 0;
    width: 100%; /* Full width */
    height: 100%; /* Full height */
    overflow: auto; /* Enable scroll if needed */
    background-color: rgb(0,0,0); /* Fallback color */
    background-color: rgba(0,0,0,0.4); /* Black w/ opacity */
}

/* Modal Content/Box */
.modal-content {
    background-color: #fefefe;
    margin: 15% auto; /* 15% from the top and centered */
    padding: 20px;
    border: 1px solid #888;
    width: 80%; /* Could be more or less, depending on screen size */
    -webkit-border-radius: 5px !important;
    -moz-border-radius: 5px !important;
    border-radius: 5px !important;
}

/* The Close Button */
.close-filePicker {
    color: #aaa;
    float: right;
    font-size: 28px;
    font-weight: bold;
}
.filePicker-title {
    color: #aaa;
    float: left;
    font-size: 28px;
    font-weight: bold;
}

.close-filePicker:hover,
.close-filePicker:focus {
    color: black;
    text-decoration: none;
    cursor: pointer;
}
    </style>
<!-- end IGV deps -->


</head>
<body>
 <div id="page-wrapper">

 	<header>
		<nav class="navbar navbar-default navbar-fixed-top" role="navigation"> <!--navbar-fixed-top-->
		  <div class="container-fluid">
		    <!-- Brand and toggle get grouped for better mobile display -->
		    <div class="navbar-header">
		      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
		        <span class="sr-only">Toggle navigation</span>
		        <span class="icon-bar"></span>
		        <span class="icon-bar"></span>
		        <span class="icon-bar"></span>
		      </button>
		      <a class="navbar-brand" href="{{ host_path }}/results">Blood eQTL browser</a>
		    </div>
		  </div><!-- /.container-fluid -->
		</nav>
 	</header>

  <h1>About this resource</h1>
  <p>This website offers visualisations of eQTL results published in (DOI goes here). Yet the primary aim of this website is to display eQTL results, the IGV genome browser offers the functionality to display also local files. Additionally, you can display your own results from a GWAS or eQTL association test in a similar way as with LocusZoom, but interactively.</p>

  <p><b>Warning:</b> The genome browser and all results are defined in reference genome <b>hg19</b>. This means that all the local files that are loaded by the user <b>must</b> be in hg19.</p>

  <h2>Display elements</h2>
  <h3>Tables</h3>
  <p>All tables can be filtered by column. Columns with numerical entries such as p-values may start with &quot;&lt;&quot;, &quot;&gt;&quot;, &quot;&lt;=&quot;, or &quot;&gt;=&quot; followed by a numerical value. Ranges are currently not supported.</p>
  <h4>Tested probes</h4>
  <p>This table displays all the probes and their gene symbols that were tested in the eQTL test. This is relevant if in the &quot;Lead variants of eQTLs&quot; there is no entry for a certain gene. As a &quot;no eQTL was found&quot; result is different from &quot;this gene has not been tested&quot;. </p>
  <h4>Lead variants of eQTLs</h4>
  <p>This table displays all eQTL associations at an FDR&lt;0.01. The eQTL test was performed using a stepwise conditional analysis - the &quot;Indep. sig.&quot; column displays the round of forward selection this association stems from; a 0 therefore indicates a primary association, a 1 the result from an association test conditioning on the association found for the same probe in iteration &quot;0&quot;, etc.</p>

  <p>On the left hand side functions for inspection of this association offered:
    <ul>
        <li><span class="glyphicon glyphicon-search" aria-hidden="true"></span> Zooms the genome browser to the location of the lead variant in the row.</li>
        <li><span class="glyphicon glyphicon-plus" aria-hidden="true"></span> Display the full summary statistic of this association in the genome browser.</li>
        <li><span class="glyphicon glyphicon-eye-open" aria-hidden="true"></span> Display transcript levels of the probe segregated by genotype.</li>
    </ul>
    If the <span class="glyphicon glyphicon-plus" aria-hidden="true"></span> is clicked the full summary statistic is displayed in the genome browser. All lead eQTL results for which the full summary statistic are displayed are highlighted in color. The color is set by cell type. 
  </p>

  <h3>Genome browser</h3>
  <p>The igv.js browser offers similar funcionality as the desktop version of IGV. Here additional functionality has been implemented to display lead eQTL results and full summary statistics in a style similar to <a href="http://locuszoom.org/">LocusZoom</a>. The genome browser here is capable of displaying local files on demand (select: &quot;Genome Browser&quot; &gt; &quot;Load local file...&quot;) to do so. Some files, such as .bam files have an index as a companion file, which is required for display and has to be defined in the file dialog.</p>

  <p>The lead eQTL results displayed in the &quot;Lead variants of eQTLs&quot; table can also be visualised in the genome browser by selecting &quot;Genome Browser&quot; &gt; &quot;&lt;cell type&gt; lead eQTLs&quot;. In the lead eQTL track displayed in the genome browser the user can click on every assocation (displayed as dot). When clicking on one of the dots click &quot;Load details&quot; which has the same effect as clicking the <span class="glyphicon glyphicon-plus" aria-hidden="true"></span> in the &quot;Lead variants of eQTLs&quot; table.</p>

  <p>Epigenetic annotation tracks for cell types can be added in order to interpret results with regard to regulatory relevant genomic regions. To add the tracks select &quot;Genome Browser&quot; &gt; &quot;Epigenetic annotation &lt;cell type&gt;&quot;</p>

  <h4>Plot your own summary statistics data</h4>
  <p>Apart from standard bioinformatics formats supported by IGV, an additional file format has been implemented to generate aforementioned LocusZoom-like visualisation from user's data. The input file for that is a plain text file, but its name must end with &quot;.signal&quot;. Its content is a tab-delimited text file in the following format:</p>

  <ul>
    <li>The file can (not required) have the following comment at the very top: &quot;#lead &lt;variantId&gt;&quot; with &lt;variantId&gt; being one of the entries in the variantId column.</li>
    <li>The file has a column header, hence it is not necessary to keep the column order and only a subset of columns can be defined.</li>
    <li>The following columns are required:</li>
    <ul>
        <li><b>chromosome</b> chromosome of the variant</li>
        <li><b>start</b> ariant start position, 1-based</li>
        <li><b>variantId</b> unique variant id, e.g: chr14:627362:A:G</li>
        <li><b>pValue</b> Nominal association p-value. If this is a result from a stepwise conditional analysis (see column &apos;fwdIter&apos;) then the p-value has to come from that analysis.</li>
        <li><b>ldR2</b> Pairwise linkeage disequilirium r<sup>2</sup> (aka coefficient of determination) between the variant in this row and the variant defined as &apos;lead&apos;. The &apos;lead&apos; variant is either defined in the file header (&apos;#lead&apos; comment) or in the column &apos;queryVarId&apos;.</li>
    </ul>
    <li>Optional additional columns that are then displayed in the info tags:</li>
    <ul>
        <li><b>snpId</b> variant id that is free to choose, preferably rs-ids</li>
        <li><b>fwdIter</b> indicate whether the lead variant indicated in the &apos;queryVarId&apos; column or in the &apos;#lead&apos; comment was selected from the nth iteration of foward-selection. Where the values in this column are then n</li>
        <li><b>geneSymbol</b> only relevant for eQTL tests: associated gene</li>
        <li><b>beta</b> effect size with respect to non-reference</li>
        <li><b>probeId</b> only relevant for array-based eQTL tests: associated microarray probe</li>
        <li><b>cellType</b> Cell type</li>
        <li><b>dataset</b> some sort of dataset identifier</li>
        <li><b>trait</b> trait name</li>
        <li><b>queryVarId</b> all entries in this column have to be the same: The lead variant with respect to which the LD is measured. The variant id in this column has to match one of the entries in the &apos;variantId&apos; column. Instead of specifying this column the &apos;#lead&apos; header comment can be used.</li>
    </ul>
    <li>Finally, if the &quot;#lead &lt;variantId&gt;&quot; is not specified in the beginning of the file, then the &quot;queryVarId&quot; column is required!</li>
  </ul>

  <h3>Presented data</h3>
  <h4>eQTL data</h4>
  <p>eQTL data from four studies was re-processed, aggregated prior to analysis. 10 PEER factors were used to account for unwanted variantion in the data.</p>

  <h4>Epigenetic annotation</h4>
  <p>The epigenetic data has been acquired from:</p>
  <ul>
    <li>Ensembl <a href="ftp://ftp.ensembl.org/pub/release-90/regulation/homo_sapiens/RegulatoryFeatureActivity/">regulatory build 90 / hg19</a>:</li>
    <ul>
        <li>Segmentation data of the cell types: CD4, CD8, CD14, CD15, CD19, MAC</li>
    </ul>
    <li>Petersen et al., 2017 (<a href="https://dx.doi.org/10.1038/ncomms16058">doi:10.1038/ncomms16058</a>)</li>
    <ul>
        <li>Segmentation data of the cell type: MK</li>
        <li>ATAC-seq data of the cell type: MK</li>
        <li>H3K27ac data of the cell type: MK</li>
    </ul>
    <li>Nuernbern et al., 2012 (<a href="https://dx.doi.org/10.1182/blood-2012-01-401893">doi:10.1182/blood-2012-01-401893</a>)</li>
    <ul>
        <li>MEIS1 ChIP-seq data of the cell type: MK</li>
    </ul>
    <li>Tijssen et al., 2011 (<a href="https://dx.doi.org/10.1016/j.devcel.2011.04.008">doi:10.1016/j.devcel.2011.04.008</a>)</li>
    <ul>
        <li>GATA1/2, RUNX1, FLI1, and SCL ChIP-seq data of the cell type: MK</li>
    </ul>
    <li>Whole-genome sequencing of rare disease patients in a national healthcare system; The NIHR BioResource, on behalf of the 100,000 Genomes Project; Submitted, 2018</li>
    <ul>
        <li>ATAC-seq data of the cell type: CD4, CD14, CD19</li>
        <li>H3K27ac data of the cell type: CD4, CD8, CD14, CD15, CD19, MAC</li>
        <li>Super enhancer data of the cell type: MK, CD4, CD8, CD14, CD15, CD19, MAC</li>
        <li>Super enhancer constituent data of the cell type: MK, CD4, CD14, CD19</li>
        <li>CTCF ChIP-seq data of the cell type: CD14, CD19</li>
    </ul>
  </ul>


  <h2>Using the REST API</h2>
  <p>REST queries can be used to download data as it is displayed. More details on this will be added soon.</p>


</div>

</body>
</html>
